### PR TITLE
Issue #468: Define initial set of design axioms.

### PR DESCRIPTION
**Pull request by pocmo**
Hey folks. Issue #468 has been around for a while and I took the time this sprint to try to define an initial set of design axioms for AC.

See those axioms as a set of generic "base rules" that define how we want to build our components and what is important to us. I do not intend to dictate those. See this PR as a proposal and that's why I will flag you all for review. Mostly I tried to capture what we have been following already... but we never wrote it down somewhere. Let me know what you think and then we can tweak as needed.

Also do not see them as "set in stone". We can adapt and complement them along the way. For example I feel like we should have something about "performance", "security" or "privacy" in here too eventually.

Here some other examples of axioms that I looked at:
- http://www.gwtproject.org/makinggwtbetter.html#designaxioms
- https://www.designprinciplesftw.com/collections/design-axioms
- http://www.mit.edu/~juhan/design_axioms.html

### CODE REVIEW

File: docs/contribute/design_axioms.md [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+
+## 💚 100% Kotlin
+
+Kotlin is the language of choice for components we develop and is the recommended language for application projects integrating components. We make use of Kotlin idioms and official Kotlin extension libraries (e.g. Coroutines).
+
+Java interoperability is not a goal of this project. We may add annotations for Java interoperability (e.g. `@JvmStatic`) at the request of a consumer to facilitate integration, but we are not considering replacing Kotlin idioms for the sake of operability.
+
+## 🔄 Third-party dependencies
+
+As providers of a framework we strive to not depend on other third-party libraries as long as not strictly needed (e.g. Android Jetpack libraries).
```
Comment (csadilek) [inline docs/contribute/design_axioms.md:15]: nit: We strive to be independent of other third-party libraries unless strictly needed...

File: docs/contribute/design_axioms.md [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+
+We are building components that directly satisfy the needs of our consumers. We tend to not design components in a vacuum without a defined use case. Components are preferred to be simple and "to the point". Additional functionality is added as needed, never in advance without knowing if it will ever be used (or how).
+
+## 🎨 Customizability
+
+Different consumers have different requirements. If possible we try to provide customization options (e.g. styling) as and if needed by consumers. We try to balance customizability and complexity. Sometimes we prefer alternative component implementations over too complex customization options.
+
+## 🧩 Pluggability
+
+Even with customization options not all components will be able to satisfy all needs of all consumers. As such, we strive to depend on interfaces (allowing different implementations) instead of classes in our code where applicable and useful. At a component level we prefer the same abstraction and try to depend on "concept components" (e.g. `concept-toolbar`) and their interfaces instead of components providing a specific implementation (e.g. `browser-toolbar`).
```
Comment (csadilek) [inline docs/contribute/design_axioms.md:31]: nit: instead of concrete classes.

File: docs/contribute/design_axioms.md [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+
+## ✅ Testing
```
Comment (csadilek) [inline docs/contribute/design_axioms.md:33]: ❤️

File: docs/contribute/design_axioms.md [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+
+We strive to have a high code coverage with a high quality suite of tests. Tests are meant to prevent regressions, exercise various parts of the code base, prove correctness and assert an always shippable state.
```
Comment (csadilek) [inline docs/contribute/design_axioms.md:35]: Suggestion:

📓 API Documentation

While we strive to keep the public API surface as simple and self-explanatory as possible, we provide KDocs for all public API methods to describe the methods purpose, parameters, return types, and possible exceptions, if applicable.

**Comment by pocmo** [comment]
> I wonder if it's worth mentioning that breaking APIs are inevitable and we don't go out of our way to avoid it (but we keep them documented).

Yes, I think that is a great idea and I totally think we should add something about API stability and guarantees we make.

File: docs/contribute/design_axioms.md (hunk 4 above)
Comment (pocmo) [reply docs/contribute/design_axioms.md:35]: That is a great addition! 👍

File: docs/contribute/design_axioms.md (hunk 1 above)
Comment (psymoon) [reply docs/contribute/design_axioms.md:15]: nit: As providers of a framework, we strive [...]

File: docs/contribute/design_axioms.md [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+
+Java interoperability is not a goal of this project. We may add annotations for Java interoperability (e.g. `@JvmStatic`) at the request of a consumer to facilitate integration, but we are not considering replacing Kotlin idioms for the sake of operability.
+
+## 🔄 Third-party dependencies
+
+As providers of a framework we strive to not depend on other third-party libraries as long as not strictly needed (e.g. Android Jetpack libraries).
+
+It is our goal to let our consumers decide what third-party libraries fit their needs, without our components making this choice or introducing an additional, duplicated stack of third-party frameworks.
+
+Additionally by avoiding excessive frameworks we intent to keep the byte size of our components, and essentially the consuming apps, as small as possible.
```
Comment (psymoon) [inline docs/contribute/design_axioms.md:19]: nit: Additionally, by limiting external dependencies, we intent to keep the byte size of our components and the consumer apps as small as possible.

File: docs/contribute/design_axioms.md [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+
+Kotlin is the language of choice for components we develop and is the recommended language for application projects integrating components. We make use of Kotlin idioms and official Kotlin extension libraries (e.g. Coroutines).
+
+Java interoperability is not a goal of this project. We may add annotations for Java interoperability (e.g. `@JvmStatic`) at the request of a consumer to facilitate integration, but we are not considering replacing Kotlin idioms for the sake of operability.
+
+## 🔄 Third-party dependencies
+
+As providers of a framework we strive to not depend on other third-party libraries as long as not strictly needed (e.g. Android Jetpack libraries).
+
+It is our goal to let our consumers decide what third-party libraries fit their needs, without our components making this choice or introducing an additional, duplicated stack of third-party frameworks.
```
Comment (psymoon) [inline docs/contribute/design_axioms.md:17]: nit: I think this is a run-on sentence.

"It is our goal to let our consumers decide what third-party libraries fit their needs; our components will not be making this choice or introducing additional and duplicated stack of third-party frameworks."

File: docs/contribute/design_axioms.md [hunk 7]
```diff
@@ -0,0 +1,35 @@
+---
+layout: page
+title: Design Axioms
+permalink: /contributing/design-axioms
+---
+
+## 💚 100% Kotlin
+
+Kotlin is the language of choice for components we develop and is the recommended language for application projects integrating components. We make use of Kotlin idioms and official Kotlin extension libraries (e.g. Coroutines).
```
Comment (psymoon) [inline docs/contribute/design_axioms.md:9]: nit: Kotlin is the language of choice for the components we develop, and it is the recommended [...]

File: docs/contribute/design_axioms.md [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+
+Additionally by avoiding excessive frameworks we intent to keep the byte size of our components, and essentially the consuming apps, as small as possible.
+
+## 🧰 Simple and applicable
+
+We are building components that directly satisfy the needs of our consumers. We tend to not design components in a vacuum without a defined use case. Components are preferred to be simple and "to the point". Additional functionality is added as needed, never in advance without knowing if it will ever be used (or how).
+
+## 🎨 Customizability
+
+Different consumers have different requirements. If possible we try to provide customization options (e.g. styling) as and if needed by consumers. We try to balance customizability and complexity. Sometimes we prefer alternative component implementations over too complex customization options.
```
Comment (psymoon) [inline docs/contribute/design_axioms.md:27]: nit: [...] "If possible, we try to provide customization options (e.g. styling) as needed by the consumers. We aim to balance customizability and complexity. When too complex of customization options are required, we prefer alternative component implementations."

File: docs/contribute/design_axioms.md (hunk 2 above)
Comment (psymoon) [reply docs/contribute/design_axioms.md:31]: nit: Even with customization options, not all components will be [...]

**Comment by pocmo** [comment]
Updated the PR with your suggestions. :)

**Comment by pocmo** [comment]
bors r=csadilek,Amejia481, jonalmeida,NotWoods,psymoon

**Pull request merged**